File: sandbox/worker.php
<?php
declare(strict_types=1);

// Queue worker: mount this directory as a shared volume and run the container with --network none.
$queue = getenv('JAVA_WORKER_QUEUE') ?: '/var/lib/java-werkstatt/queue';
$java = getenv('JAVA_BIN') ?: 'java';
$poll = 20000;

foreach (['in', 'out'] as $directory) {
    $path = $queue . DIRECTORY_SEPARATOR . $directory;
    if (!is_dir($path)) mkdir($path, 0700, true);
}

while (true) {
    $jobs = glob($queue . '/in/*/READY') ?: [];
    foreach ($jobs as $ready) processJob(dirname($ready), $queue . '/out', $java);
    usleep($poll);
}

function processJob(string $jobDir, string $outputDir, string $java): void
{
    $claim = $jobDir . '/RUNNING';
    if (!@rename($jobDir . '/READY', $claim)) return;
    $job = json_decode((string) @file_get_contents($jobDir . '/job.json'), true);
    $jobId = basename($jobDir);
    if (!is_array($job) || !preg_match('/^[A-Za-z_$][A-Za-z0-9_$]*$/', (string) ($job['mainClass'] ?? ''))) {
        writeResult($outputDir, $jobId, ['exitCode' => 2, 'stdout' => '', 'stderr' => 'Trabajo inválido.', 'timedOut' => false]);
        removeDirectory($jobDir);
        return;
    }
    $command = [$java, '-Duser.language=en', '-Duser.country=US', '-Dfile.encoding=UTF-8', '-Djava.awt.headless=true', '-Xms16m', '-Xmx64m', '-XX:ActiveProcessorCount=1', '-cp', $jobDir . '/classes', $job['mainClass']];
    $result = runProcess($command, $jobDir, (float) (getenv('JAVA_WORKER_TIMEOUT') ?: 3), (string) ($job['stdin'] ?? ''));
    writeResult($outputDir, $jobId, $result);
    removeDirectory($jobDir);
}

function runProcess(array $command, string $cwd, float $timeout, string $stdin): array
{
    $pipes = [];
    $process = proc_open($command, [0 => ['pipe', 'r'], 1 => ['pipe', 'w'], 2 => ['pipe', 'w']], $pipes, $cwd, ['LC_ALL' => 'C', 'LANG' => 'C']);
    if (!is_resource($process)) return ['exitCode' => 1, 'stdout' => '', 'stderr' => 'No se pudo iniciar Java.', 'timedOut' => false];
    fwrite($pipes[0], $stdin); fclose($pipes[0]);
    stream_set_blocking($pipes[1], false); stream_set_blocking($pipes[2], false);
    $started = microtime(true); $stdout = ''; $stderr = ''; $timedOut = false;
    do {
        $stdout .= stream_get_contents($pipes[1]) ?: '';
        $stderr .= stream_get_contents($pipes[2]) ?: '';
        $status = proc_get_status($process);
        if (!$status['running']) break;
        if (microtime(true) - $started > $timeout) {
            $timedOut = true; proc_terminate($process); usleep(100000); proc_terminate($process, 9); break;
        }
        usleep(20000);
    } while (strlen($stdout) + strlen($stderr) < 12000);
    $stdout .= stream_get_contents($pipes[1]) ?: ''; $stderr .= stream_get_contents($pipes[2]) ?: '';
    fclose($pipes[1]); fclose($pipes[2]);
    return ['exitCode' => proc_close($process), 'stdout' => substr($stdout, 0, 12000), 'stderr' => substr($stderr, 0, 12000), 'timedOut' => $timedOut];
}

function writeResult(string $outputDir, string $jobId, array $result): void
{
    file_put_contents($outputDir . '/' . $jobId . '.tmp', json_encode($result, JSON_UNESCAPED_UNICODE), LOCK_EX);
    @rename($outputDir . '/' . $jobId . '.tmp', $outputDir . '/' . $jobId . '.json');
}

function removeDirectory(string $directory): void
{
    if (!is_dir($directory)) return;
    foreach (scandir($directory) ?: [] as $entry) {
        if ($entry === '.' || $entry === '..') continue;
        $path = $directory . '/' . $entry;
        is_dir($path) ? removeDirectory($path) : @unlink($path);
    }
    @rmdir($directory);
}

File: sandbox/Worker.java
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.AtomicMoveNotSupportedException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.time.Duration;
import java.util.Comparator;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.regex.Pattern;

final class Worker {
    private static final Path QUEUE = Path.of(System.getenv().getOrDefault(
        "JAVA_WORKER_QUEUE", "/var/lib/java-werkstatt/queue"
    ));
    private static final Duration COMPILE_TIMEOUT = Duration.ofSeconds(8);
    private static final Duration RUN_TIMEOUT = Duration.ofSeconds(3);
    private static final int MAX_OUTPUT_BYTES = 12_000;
    private static final Pattern MAIN_CLASS = Pattern.compile("[A-Za-z_$][A-Za-z0-9_$]*");
    private static final Pattern FILE_NAME = Pattern.compile("[A-Za-z_$][A-Za-z0-9_$]*\\.java");

    public static void main(String[] args) throws Exception {
        Files.createDirectories(QUEUE.resolve("in"));
        Files.createDirectories(QUEUE.resolve("out"));
        while (!Thread.currentThread().isInterrupted()) {
            try (var jobs = Files.list(QUEUE.resolve("in"))) {
                jobs.filter(Files::isDirectory).forEach(Worker::processIfReady);
            }
            Thread.sleep(20);
        }
    }

    private static void processIfReady(Path jobDirectory) {
        Path ready = jobDirectory.resolve("READY");
        Path running = jobDirectory.resolve("RUNNING");
        if (!Files.isRegularFile(ready)) return;
        try {
            Files.move(ready, running, StandardCopyOption.ATOMIC_MOVE);
        } catch (IOException exception) {
            return;
        }

        String jobId = jobDirectory.getFileName().toString();
        try {
            String mainClass = Files.readString(jobDirectory.resolve("main-class.txt")).trim();
            String fileName = Files.readString(jobDirectory.resolve("file-name.txt")).trim();
            boolean shouldRun = Files.readString(jobDirectory.resolve("run.txt")).trim().equals("1");
            if (!MAIN_CLASS.matcher(mainClass).matches()) {
                throw new IllegalArgumentException("Clase principal inválida.");
            }
            if (!FILE_NAME.matcher(fileName).matches()) {
                throw new IllegalArgumentException("Nombre de archivo inválido.");
            }
            byte[] stdin = Files.readAllBytes(jobDirectory.resolve("stdin.bin"));
            Result result = compileAndRun(jobDirectory, fileName, mainClass, stdin, shouldRun);
            writeResult(jobId, result);
        } catch (Exception exception) {
            writeResult(jobId, new Result("worker", 2, "", exception.getMessage(), "", false, 0, 0));
        } finally {
            removeDirectory(jobDirectory);
        }
    }

    private static Result compileAndRun(
        Path jobDirectory,
        String fileName,
        String mainClass,
        byte[] stdin,
        boolean shouldRun
    ) throws Exception {
        Path classes = jobDirectory.resolve("classes");
        Files.createDirectory(classes);
        ProcessResult compilation = execute(List.of(
            "javac", "-J-Duser.language=en", "-J-Duser.country=US", "-encoding", "UTF-8",
            "-J-Xmx96m", "-proc:none", "-Xlint:all", "-d", classes.toString(),
            jobDirectory.resolve(fileName).toString()
        ), jobDirectory, COMPILE_TIMEOUT, new byte[0]);
        String compileOutput = (compilation.stdout() + "\n" + compilation.stderr())
            .replace(jobDirectory.toString() + "/", "")
            .trim();
        if (compilation.exitCode() != 0 || compilation.timedOut() || !shouldRun) {
            return new Result(
                "compile",
                compilation.exitCode(),
                "",
                "",
                compileOutput,
                compilation.timedOut(),
                compilation.durationMs(),
                0
            );
        }

        ProcessResult execution = execute(List.of(
            "java", "-Duser.language=en", "-Duser.country=US", "-Dfile.encoding=UTF-8",
            "-Djava.awt.headless=true", "-Xms16m", "-Xmx64m", "-XX:ActiveProcessorCount=1",
            "-cp", classes.toString(), mainClass
        ), jobDirectory, RUN_TIMEOUT, stdin);
        return new Result(
            "run",
            execution.exitCode(),
            execution.stdout(),
            execution.stderr(),
            compileOutput,
            execution.timedOut(),
            compilation.durationMs(),
            execution.durationMs()
        );
    }

    private static ProcessResult execute(
        List<String> command,
        Path jobDirectory,
        Duration timeout,
        byte[] stdin
    ) throws Exception {
        long started = System.nanoTime();
        Process process = new ProcessBuilder(command).directory(jobDirectory.toFile()).start();
        try (var input = process.getOutputStream()) {
            input.write(stdin);
        }
        CompletableFuture<String> stdout = readLimited(process.getInputStream());
        CompletableFuture<String> stderr = readLimited(process.getErrorStream());
        boolean finished = process.waitFor(timeout.toMillis(), TimeUnit.MILLISECONDS);
        if (!finished) {
            process.destroy();
            if (!process.waitFor(100, TimeUnit.MILLISECONDS)) process.destroyForcibly();
        }
        return new ProcessResult(
            finished ? process.exitValue() : 124,
            stdout.get(1, TimeUnit.SECONDS),
            stderr.get(1, TimeUnit.SECONDS),
            !finished,
            TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - started)
        );
    }

    private static CompletableFuture<String> readLimited(InputStream stream) {
        return CompletableFuture.supplyAsync(() -> {
            try (stream; var output = new ByteArrayOutputStream()) {
                byte[] buffer = new byte[1024];
                int total = 0;
                for (int read; (read = stream.read(buffer)) != -1;) {
                    int accepted = Math.min(read, MAX_OUTPUT_BYTES - total);
                    if (accepted > 0) output.write(buffer, 0, accepted);
                    total += accepted;
                }
                String text = output.toString(StandardCharsets.UTF_8);
                return total > MAX_OUTPUT_BYTES ? text + "\n… salida recortada …" : text;
            } catch (IOException exception) {
                return "";
            }
        });
    }

    private static void writeResult(String jobId, Result result) {
        Path temporary = QUEUE.resolve("out").resolve(jobId + ".tmp");
        Path destination = QUEUE.resolve("out").resolve(jobId + ".json");
        String json = """
            {"phase":"%s","exitCode":%d,"stdout":"%s","stderr":"%s","compileOutput":"%s","timedOut":%s,"durationMs":%d,"compileDurationMs":%d,"runDurationMs":%d}
            """.formatted(
                result.phase(),
                result.exitCode(),
                escape(result.stdout()),
                escape(result.stderr()),
                escape(result.compileOutput()),
                result.timedOut(),
                result.compileDurationMs() + result.runDurationMs(),
                result.compileDurationMs(),
                result.runDurationMs()
            ).trim();
        try {
            Files.writeString(temporary, json, StandardCharsets.UTF_8);
            try {
                Files.move(temporary, destination, StandardCopyOption.ATOMIC_MOVE);
            } catch (AtomicMoveNotSupportedException exception) {
                Files.move(temporary, destination, StandardCopyOption.REPLACE_EXISTING);
            }
        } catch (IOException exception) {
            System.err.println("No se pudo escribir el resultado " + jobId + ": " + exception.getMessage());
        }
    }

    private static String escape(String value) {
        return value.replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\r", "\\r")
            .replace("\n", "\\n")
            .replace("\t", "\\t");
    }

    private static void removeDirectory(Path directory) {
        if (!Files.exists(directory)) return;
        try (var paths = Files.walk(directory)) {
            paths.sorted(Comparator.reverseOrder()).forEach(path -> {
                try {
                    Files.deleteIfExists(path);
                } catch (IOException ignored) {
                }
            });
        } catch (IOException ignored) {
        }
    }

    private record ProcessResult(
        int exitCode,
        String stdout,
        String stderr,
        boolean timedOut,
        long durationMs
    ) {
    }

    private record Result(
        String phase,
        int exitCode,
        String stdout,
        String stderr,
        String compileOutput,
        boolean timedOut,
        long compileDurationMs,
        long runDurationMs
    ) {
    }
}
